### PR TITLE
feat(types) Remove dependency to `cranelift-entity`

### PR DESCRIPTION
**Pull request by Hywan**
# Description

Don't be afraid, everything is cool.

`cranelift-entity` is nice, but we are limited by it. We want to go
further for some features. Those features are very likely to be
refused by Cranelift because it's out of their scope.

This patch basically copies the code from `cranelift-entity` inside
`wasmer-types`. The code has been modified to fit in our own
“`no_std`” strategy. The code will be modified deeper in upcoming
patches and most of the entities are likely to radically change. The
goal of this patch is strictly to not break our code while porting the
new code.

# Review

- [ ] Add a short description of the the change to the CHANGELOG.md file
- [ ] Do we need to update the `ATTRIBUTIONS.md` file?

### CODE REVIEW

**Comment by Hywan** [comment]
bors try

**Comment by Hywan** [comment]
bors try

File: lib/wasmer-types/src/entity/mod.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                f: &mut $crate::lib::std::fmt::Formatter,
+            ) -> $crate::lib::std::fmt::Result {
+                write!(f, concat!($display_prefix, "{}"), self.0)
+            }
+        }
+
+        impl $crate::lib::std::fmt::Debug for $entity {
+            fn fmt(
+                &self,
+                f: &mut $crate::lib::std::fmt::Formatter,
```
Comment (MarkMcCaskey) [inline lib/wasmer-types/src/entity/mod.rs:71]: Does this macro expand to the right thing in other crates?

File: lib/wasmer-types/src/entity/packed_option.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
+//! Compact representation of `Option<T>` for types with a reserved value.
+//!
+//! Small Cranelift types like the 32-bit entity references are often used in tables and linked
```
Comment (MarkMcCaskey) [inline lib/wasmer-types/src/entity/packed_option.rs:3]: Do we want to generalize the doc comments? We use these types for more than just interacting with Cranelift.

File: lib/wasmer-types/src/entity/packed_option.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+//! to represent `None`.
+
+use crate::lib::std::fmt;
+use crate::lib::std::mem;
+
+#[cfg(feature = "enable-serde")]
+use serde::{Deserialize, Serialize};
+
+/// Types that have a reserved value which can't be created any other way.
+pub trait ReservedValue {
```
Comment (MarkMcCaskey) [inline lib/wasmer-types/src/entity/packed_option.rs:17]: Do we use any of the types in this module in Wasmer? If not, perhaps we can just not include them

File: lib/wasmer-types/src/entity/secondary_map.rs [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        }
+    }
+
+    /// Returns the number of elements the map can hold without reallocating.
+    pub fn capacity(&self) -> usize {
+        self.elems.capacity()
+    }
+
+    /// Get the element at `k` if it exists.
+    #[inline(always)]
```
Comment (MarkMcCaskey) [inline lib/wasmer-types/src/entity/secondary_map.rs:86]: A bit odd to see these on `SecondaryMap` and not on `PrimaryMap` but 🤷 . I find it easy to believe that these should always be inlined and that Rust probably also agrees and doesn't need to be told.

Though I think the `#[inline]` attributes _do_ matter for cross-crate inlining without LTO? With LTO they don't matter, but without LTO they do. At least that's what I remember, not super confident in that memory though.

Anyways, no action needed here. Just pointing it out.

**Review by MarkMcCaskey — APPROVED** [review]
Looks good to me.

👍 on the attributions. I'm in favor of being explicit about where the code came from and what the license is

File: lib/wasmer-types/src/entity/mod.rs (hunk 1 above)
Comment (Hywan) [reply lib/wasmer-types/src/entity/mod.rs:71]: Yes. The `lib` module is now public. `wasmer-compiler-llvm` uses this macro, and `cargo check` + tests are OK.

File: lib/wasmer-types/src/entity/packed_option.rs (hunk 2 above)
Comment (Hywan) [reply lib/wasmer-types/src/entity/packed_option.rs:3]: :+1:

File: lib/wasmer-types/src/entity/packed_option.rs (hunk 3 above)
Comment (Hywan) [reply lib/wasmer-types/src/entity/packed_option.rs:17]: Will check. I've already removed some types that we aren't using.

File: lib/wasmer-types/src/entity/secondary_map.rs (hunk 4 above)
Comment (Hywan) [reply lib/wasmer-types/src/entity/secondary_map.rs:86]: Yes. That's part of the rewrite I want to do later. Let's keep the same code for the moment, and let's optimize later as planned :-). Thanks for pointing that out!

**Comment by syrusakbary** [comment]
Note: on the attributions we should follow our current structure. Which is:
* No need to change `ATTRIBUTIONS.md` since the following is still true: `wasmtime: for their API and internal documentation, as well as some internal implementations`
* [ ] We need to update [`wasmer-types/README.md`](https://github.com/wasmerio/wasmer/blob/master/lib/wasmer-types/README.md) adding an Acknowledgments section with the cranelift-entity reference.
```
### Acknowledgments

This project borrowed some of the code for the entity structure from [cranelift-entity](https://crates.io/crates/cranelift-entity).
We decided to move it here to help on serialization/deserialization and also to ease the integration with other tools like loupe.

Please check [Wasmer ATTRIBUTIONS](https://github.com/wasmerio/wasmer/blob/master/ATTRIBUTIONS.md) to further see licenses and other attributions of the project. 
```

* [ ] We need the following content to the headers in the new added files:

```
// This file contains code from external sources.
// Attributions: https://github.com/wasmerio/wasmer/blob/master/ATTRIBUTIONS.md

```

File: lib/wasmer-types/src/entity/packed_option.rs (hunk 3 above)
Comment (Hywan) [reply lib/wasmer-types/src/entity/packed_option.rs:17]: I confirmed it's used.

**Pull request merged**